File: LeetCode/Easy/594-Longest_Harmonious_Subsequence/594-LongestHarmoniousSubsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findLHS = function (nums) {
    let freq = new Map();

    for (const num of nums) {
        freq.set(num, (freq.get(num) || 0) + 1);
    }

    let ans = 0;

    for (const [k, v] of freq) {
        if (freq.has(k + 1)) {
            ans = ans > v + freq.get(k + 1) ? ans : v + freq.get(k + 1);
        }
    }

    return ans;
};

File: LeetCode/Easy/594-Longest_Harmonious_Subsequence/594-LongestHarmoniousSubsequence.py
from collections import Counter


class Solution:
    def findLHS(self, nums: list[int]) -> int:
        freq = Counter(nums)
        ans = 0

        for k, v in freq.items():
            if freq[k + 1] != 0:
                ans = ans if ans > freq[k] + freq[k + 1] else freq[k] + freq[k + 1]

        return ans

    def findLHS1(self, nums: list[int]) -> int:
        nums.sort()
        n = len(nums)

        ans = 0
        j = 0

        for i in range(n):
            while j < n and nums[j] - nums[i] <= 1:
                j += 1

            if nums[i] != nums[j - 1]:
                ans = max(ans, j - i)

        return ans


a = Solution()
print(a.findLHS(nums=[1, 3, 2, 2, 5, 2, 3, 7]))
print(a.findLHS(nums=[1, 2, 3, 4]))
print(a.findLHS(nums=[1, 1, 1, 1]))
